src/diffusers/models/attention.py

```diff
@@ -214,6 +239,53 @@ def forward(self, hidden_states, context=None):
         return hidden_states
 
 
+class MemoryEfficientCrossAttention(nn.Module):
+    def __init__(self, query_dim, context_dim=None, heads=8, dim_head=64, dropout=0.0):
+        super().__init__()
+        inner_dim = dim_head * heads
+        context_dim = default(context_dim, query_dim)
+
+        self.heads = heads
+        self.dim_head = dim_head
+
+        self.to_q = nn.Linear(query_dim, inner_dim, bias=False)
+        self.to_k = nn.Linear(context_dim, inner_dim, bias=False)
+        self.to_v = nn.Linear(context_dim, inner_dim, bias=False)
+
+        self.to_out = nn.Sequential(nn.Linear(inner_dim, query_dim), nn.Dropout(dropout))
+        self.attention_op: Optional[Any] = None
+
+    def forward(self, x, context=None, mask=None):
+        q = self.to_q(x)
+        context = default(context, x)
+        k = self.to_k(context)
+        v = self.to_v(context)
+
+        b, _, _ = q.shape
+        q, k, v = map(
+            lambda t: t.unsqueeze(3)
+            .reshape(b, t.shape[1], self.heads, self.dim_head)
+            .permute(0, 2, 1, 3)
+            .reshape(b * self.heads, t.shape[1], self.dim_head)
+            .contiguous(),
+            (q, k, v),
+        )
+
+        # actually compute the attention, what we cannot get enough of
+        out = xformers.ops.memory_efficient_attention(q, k, v, attn_bias=None, op=self.attention_op)
```

Comment:
@MatthieuTPHR do you happen to know an answer to my last question

> I also noticed in my experiment that we don't use Flash Attention, while in [HazyResearch's implem](https://github.com/HazyResearch/flash-attention) we do (for example when self.dim_head=40). Are they different implementations, or should we fix the conditions to use FlashAttention? ([original conds](https://github.com/HazyResearch/diffusers/blob/e5017505fc1c0b6725dcb67396d6af2ac51d5b6e/src/diffusers/models/attention.py#L304-L305) vs [conds in xformers](https://github.com/facebookresearch/xformers/blob/51dd1192620271598ba71bb1351c1c5508720b31/xformers/ops.py#L300))

Would it be worth it to consider a `FlashCrossAttention` in the future as well to use HazyResearch's Flash Attention? Or is it the same as the one in `xformers`?

Reply:
That's a good point!
xFormers supports both Flash & our own kernels as backends.
With `attention_op = None`, xFormers will try to infer which one is (1) supported (2) the fastest. If you want to force using a specific backend, you can set `op = xformers.ops.MemoryEfficientAttentionFlashAttentionOp` for instance (which we don't recommend, because when we optimise a kernel for a specific case, we can just update the dispatcher, and users with `op=None` will benefit from the improvements right away).

> Are they different implementations, or should we fix the conditions to use FlashAttention? ([original conds](https://github.com/HazyResearch/diffusers/blob/e5017505fc1c0b6725dcb67396d6af2ac51d5b6e/src/diffusers/models/attention.py#L304-L305) vs [conds in xformers](https://github.com/facebookresearch/xformers/blob/51dd1192620271598ba71bb1351c1c5508720b31/xformers/ops.py#L300))

That's a good catch. Upon until [a few days ago](https://github.com/HazyResearch/flash-attention/commit/46fd2a20b20849598b1abb438ea09520449d7eb3), Flash only supported a few values of K. They extended the coverage and made a few improvements - I plan to update xFormers with the new version (and new conditions) this week or next week.